### ISSUE TITLE
CUDA out of memory while processing long tracks

### ISSUE THREAD

**Issue by famzah** [facebookresearch/demucs#231]
Hello,

First of all -- thank you for this great product. It works flawlessly using CPU.

I'm trying to process material faster by using a GPU on an AWS EC2 instance. Unfortunately, it terminates with the following error:
```bash
$ demucs audio.mp3 -d cuda

Selected model is a bag of 4 models. You will see that many progress bars per track.
Separated tracks will be stored in /home/ec2-user/separated/mdx_extra_q
Separating track audio.mp3
100%|██████████████████████████████████████████████████████████████████████| 4356.0/4356.0 [01:35<00:00, 45.42seconds/s]

Traceback (most recent call last):
  File "/home/ec2-user/.local/bin/demucs", line 8, in <module>
    sys.exit(main())
  File "/home/ec2-user/.local/lib/python3.7/site-packages/demucs/separate.py", line 120, in main
    overlap=args.overlap, progress=True)[0]
  File "/home/ec2-user/.local/lib/python3.7/site-packages/demucs/apply.py", line 147, in apply_model
    estimates += out
RuntimeError: CUDA out of memory. Tried to allocate 5.71 GiB (GPU 0; 14.76 GiB total capacity; 9.12 GiB already allocated; 4.33 GiB free; 9.15 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation.  See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF
```

Info about the running environment:
- Python 3.7.10 and PyTorch 1.10.0
- AWS EC2 instance type: g4dn.xlarge
- Operating system and version:
```
Deep Learning AMI GPU PyTorch 1.10.0 (Amazon Linux 2) 20211115
```
- GPU Hardware:
```
NVIDIA Corporation TU104GL [Tesla T4] (rev a1)
Driver Version: 470.57.02    CUDA Version: 11.4
15109MiB memory
```

Do the models require a card with more than 16 GB RAM? If that's not your experience, can you share your hardware/software environment, so that I can retry. Thank you.

**Comment by melonjack** [comment]
Almost same here. After first install all was flawless, but on the second pc that happens. Had to create conda env again on first one - seems like same steps was maded, but encounter ssme problem((((

**Comment by lucellent** [comment]
It's an MDX limitation as far as I know, I have 3070 with 8GB VRAM and 64GB RAM and still get that issue with tracks longer than 4 minutes.

**Comment by famzah** [comment]
@adefossez, can you share the specs of your hardware and software environment where you test the CUDA implementation?

It's a fact that I'm running this on a fairly long audio track (1+ hour), so @lucellent is giving a valuable hint.

**Comment by lucellent** [comment]
Dang, 1 hour is way, way, way too long. Even with my hardware, as I mentioned, my limit is about 4 minutes, and I don't think it's that much hardware dependant. The new demucs v3 is based on MDX (which is another tool similar to demucs) but it has those memory issues, that people haven't been able to fix yet, if it's even possible

You can either chop up your song (1 hour is a lot, so you will have to chop it many times), or do CPU conversion

**Comment by famzah** [comment]
I can definitely chop it automatically even at 1-minute lengths but I guess this will affect the quality and there will be clicks where the song is split.

The strange thing is that the long separation process finishes successfully as shown in the first post of this issue thread. Therefore, the hard work is already done. The error occurs at the very final stage (line 147, in apply_model). If we could do just this "apply_model" step using CPU since CUDA is having memory issues... :)

**Comment by lucellent** [comment]
I actually have already tried the chopping method and as far as I checked, there weren't any clicks or anything between tracks. But you can always add a few seconds just in case something like this happens, but you'd have to align all the tracks yourself after that (since they will have a few seconds before and after the tracks just for the margin of error)

About the separation process - I don't think it's nowhere near done. Demucs will show a few bars PER stem, so you'd have a few bars (say 4) that will show 4 times (because demucs outputs 4 stems). I can only see one bar on your post.

**Comment by lucellent** [comment]
If you feel like it, I might give a shot at your track too, or with CPU (without chopping). But I'm just suggesting, feel free to try it on your own again :)

**Comment by famzah** [comment]
In the source code I see that "demucs" is already working in chunks of 8 seconds:
```
def apply_model():

    Apply model to a given mixture.

    Args:
        shifts (int): if > 0, will shift in time `mix` by a random amount between 0 and 0.5 sec
            and apply the oppositve shift to the output. This is repeated `shifts` time and
            all predictions are averaged. This effectively makes the model time equivariant
            and improves SDR by up to 0.2 points.
        split (bool): if True, the input will be broken down in 8 seconds extracts
            and predictions will be performed individually on each and concatenated.
            Useful for model with large memory footprint like Tasnet.
```

So splitting a long audio track into many chunks seems a valid approach.

The only question is, do I need to split into overlapping parts, so that the algo has enough extra time to achieve best results?

Let's imagine that we mark each 1 minute of length with consecutive letters and I have the following track: "ABCDE". If I split the track with no overlap, I get "A", "B", "C", "D", "E". But if I split with overlap, I would get "ABC", "BCD", "CDE", etc. Once processed by "demucs", I can cut the extra overlap:
- "ABC" becomes "B"
- "BCD" becomes "C"
- "CDE" becomes "D"

Finally, I concatenate the processed parts "A", "B", "C", "D", "E" into a track with the original timing "ABCDE".

@adefossez, is such overlap necessary for the algorithm to produce the best results? Or I can simply cut the long track, process each part and then concatenate the parts into a long track again as suggested by @lucellent?

**Comment by famzah** [comment]
@adefossez, I actually dug myself into the source code of "demucs" and found out that you're already doing the same thing in regards to split and overlaps. So I spent my efforts into making "demucs" more memory efficient, rather than inventing my own wheel.

I've got a working PoC which reduces the GPU memory usage to 1.3 GB (which is the size of the model) and no other GPU memory is used regardless of the length of the track. Any track requires just 1.3 GB of GPU memory as reported by "torch".

The cost for this optimization is about 7% slower performance and of course more "traditional" RAM usage. Still, a machine with plenty of RAM is easy to get, while the GPU memory is very limited. The CPU usage increases with 10-15% but it's overall amount is very low anyway.

Would you accept this as a permanent solution for "demucs"? Lose 7% performance but be able to process arbitrary long tracks, and at the same time reduce the GPU memory requirement of "demucs" to only 2 GB VRAM?

**Comment by adefossez** [comment]
Sorry not replying much I'm on vacation. @famzah you can open a pull request and i will have a look.

I had ideas to further reduce the memory foot print (in particular keeping the model quantized until we use it).

Other sources of memory usage are storing the full stems in float32. This can be quickly non negligible (1h of audio can be up to 7GB I think), so for long tracks (20min) this can add 1 to 2 GB.

We could make apply_model actually become an iterator that yields chunks of audio for 1 min and save on disk streaming. For bags of model though this require cycling through the models for each chunks, while currently I cycle only once.

So yeah, there is a lot of room for improvement but this is likely going to make the apply_model more complex and the changes are non trivial.

**Comment by BitBlueDevStudio** [comment]
Hello, I have similliar problem on cpu. If I process track around 50 seconds, it takes around 3gb of RAM.
If I process track with length around 5-7 minutes, it takes from 3gb to 5gb RAM while processing.

So if I run 3 parallel processes with 12gb RAM computer, on of these 3 processes will crash with "unable to allocate memory", because at some point 12gb RAM is not enough for one of 3 processes.

I don't sure, but it looks a little bit weird, when you processing 3 song each size around 5-10mb :))))

Is that normal behavior or we have here some problem? Any hot fix available?

Thanks in advance!

**Comment by mepc36** [comment]
@famzah what AWS EC2 type are you using? I'm trying to implement the same exact architecture with demucs, and running into similar problems. I'm using `t2.large`, thanks!

**Comment by famzah** [comment]
@mepc36, I'm using a g4dn.2xlarge AWS EC2 instance (32 GB RAM, GPU with 16 GB VRAM). I would recommend that you try with a much larger instance and scale down until you've found what works for your particular workload.

@BitBlueDevStudio, I don't know about the CPU use-case but for CUDA only the AI model takes 1.3 GB GPU VRAM. So, yes, some base RAM usage is expected regardless of how long your track is.

**Comment by famzah** [comment]
@adefossez, enjoy your vacation. This subject is not urgent at all.

> We could make apply_model actually become an iterator that yields chunks of audio for 1 min and save on disk streaming. For bags of model though this require cycling through the models for each chunks, while currently I cycle only once.

This idea struck me, too, while I was reviewing the code. Unfortunately, we still need to post-process the whole output, in order to be able to normalize it. So while working in chunks would be memory efficient, we would need to save the raw data chunks on disk, remember the peak value, then read back all raw data chunks, normalize them and finally save as a WAV file.

Working in chunks is suitable also for the input data. Right now the whole input audio file is read in memory while we only need a tiny chunk from it at a time. Even with the current implementation of apply_model().

Overall, the idea for an iterator based processing is efficient but seems complicated to be implemented. As long as I can compensate this inefficiency with regular RAM (not GPU VRAM), I consider this optimization a low priority feature.

**Comment by famzah** [comment]
When you review PR #234, I will submit a PR for the code which uses much less GPU VRAM at the expense of more regular RAM.

**Comment by yuukiyuuna** [comment]
> When you review PR #234, I will submit a PR for the code which uses much less GPU VRAM at the expense of more regular RAM.

我赞同这个做法
这个分支可以让我参考一下么

**Comment by JohnSmith2007** [comment]
Having the same issue with the Collab, except my audio file is only 30 minutes

**Comment by yuukiyuuna** [comment]
> Having the same issue with the Collab, except my audio file is only 30 minutes

Hi, according to your modification, the memory problem has been successfully solved, but this has caused new problems. I also want to ask for some advice, because I am not too familiar with this piece.

The video size is about 30M

```
(demucs) PS E:\test> demucs -d cuda .\linshi-mv.mp4
Selected model is a bag of 4 models. You will see that many progress bars per track.
Separated tracks will be stored in E:\test\separated\mdx_extra_q
Separating track linshi-mv.mp4
0%| | 0.0/1782.0 [00:00<?, ?seconds/s]Traceback (most recent call last):
File "C:\ProgramData\Miniconda3\envs\demucs\Scripts\demucs-script.py", line 33, in
sys.exit(load_entry_point('demucs', 'console_scripts', 'demucs')())
File "e:\project\demucs\demucs\separate.py", line 120, in main
overlap=args.overlap, progress=True)[0]
File "e:\project\demucs\demucs\apply.py", line 143, in apply_model
transition_power=transition_power, progress=progress)
File "e:\project\demucs\demucs\apply.py", line 175, in apply_model
chunk_out = apply_model(model, chunk, shifts=shifts, split=False)
File "e:\project\demucs\demucs\apply.py", line 191, in apply_model
shifted_out = apply_model(model, shifted, shifts=0, split=False)
File "e:\project\demucs\demucs\apply.py", line 203, in apply_model
out = model(padded_mix)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "e:\project\demucs\demucs\hdemucs.py", line 711, in forward
x = encode(x, inject)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "e:\project\demucs\demucs\hdemucs.py", line 128, in forward
y = self.dconv(y)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "e:\project\demucs\demucs\demucs.py", line 165, in forward
x = x + layer(x)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\container.py", line 141, in forward
input = module(input)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "e:\project\demucs\demucs\demucs.py", line 48, in forward
x = self.linear(x)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\module.py", line 1102, in _call_impl
return forward_call(*input, **kwargs)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\modules\linear.py", line 103, in forward
return F.linear(input, self.weight, self.bias)
File "C:\ProgramData\Miniconda3\envs\demucs\lib\site-packages\torch\nn\functional.py", line 1848, in linear
return torch._C._nn.linear(input, weight, bias)
RuntimeError: CUDA error: CUBLAS_STATUS_NOT_INITIALIZED when calling cublasCreate(handle)
0%| | 0.0/1782.0 [00:00<?, ?seconds/s]
```

I checked it may be because of the batch size setting, but I don't know how to avoid it.
Hope to give some guidance，Thanks

**Comment by famzah** [comment]
@yuukiyuuna, PR #234 does not solve the memory issues. It's a simple optimization which should not cause any problems. I guess that without PR #234 you will get the same error. Please open another issue, because it's not related to the memory problems that we discuss here.

**Comment by famzah** [comment]
@adefossez, I've submitted PR #244 for the code which uses much less GPU VRAM at the expense of more regular RAM. Please review and if there's anything to discuss, let's comment in the PR.

**Comment by mepc36** [comment]
> @adefossez, I've submitted PR #244 for the code which uses much less GPU VRAM at the expense of more regular RAM. Please review and if there's anything to discuss, let's comment in the PR.

This is AWESOME help @famzah, thanks for writing that PR! I would've helped but I don't know Python super well haha...was reading your benchmarks and those tradeoffs b/w memory and processing time are definitely ones I'd sign up for

**Comment by adefossez** [comment]
Thanks a lot @famzah  I have integrated your changes into this PR https://github.com/facebookresearch/demucs/pull/248, which also includes moving the models one by one to the GPU and out of it which saves a bit extra space, although not that much (now just under 5GB).

I'll be merging this into main soon, once all the tests have run.

There could be a bit more savings to be done, although it would be a bit technical. Innermost layers can probably be evaluated as float16, but this requires more careful experimentation to make sure it does not impact quality.

**Comment by famzah** [comment]
@adefossez, I'm glad that you were inspired to load the sub-models one by one to the GPU, too. At this point I consider the optimizations enough for practical use with arbitrary long audio tracks. The limited amount of GPU memory is now used efficiently and the rest goes into the regular CPU RAM of which we can provision tens of GBytes, if we have to.

There is one small thing to consider which I [commented](https://github.com/facebookresearch/demucs/pull/248#issuecomment-991250340) in the PR.

Next week I should have the opportunity to try some very longs tracks using the new code. I will then report back here, and close this issue if everything works correctly.

Thank you for all your work.

**Comment by adefossez** [comment]
as mentioned i nthe PR, I left the initial loading on CUDA because some resampling might happen and I wanted it to happen on GPU. not sure how long resampling 1 hour of audio can take on CPU though.

**Comment by famzah** [comment]
Yeap. I will test this week and report back.

**Comment by famzah** [comment]
The changes in PR #248 fixed the memory issues! Thank you.

I've tested with a track which is 136 minutes long.

The maximum memory usage reported by "nvidia-smi" was 5849 MiB. The actually used GPU memory is much less, as reported by [torch.cuda.memory_allocated()](https://pytorch.org/docs/stable/generated/torch.cuda.memory_allocated.html#torch.cuda.memory_allocated). PyTorch [memory management](https://pytorch.org/docs/stable/notes/cuda.html#cuda-memory-management) is complicated. It may be that it allocates big memory (cache) blocks, or it could be something else. I see no reason to investigate this further because GPUs with 6 GB VRAM are widely available.

The peak of the used regular RAM was 44 GB. That's a lot but machines with 64 GB or more are easy to provision nowadays.

---

Here are the raw results:
```
$ /bin/time -v demucs -d cuda track6.wav 
Selected model is a bag of 4 models. You will see that many progress bars per track.
Separated tracks will be stored in /home/volatile/famzah/separated/mdx_extra_q
Separating track track6.wav
100%|██████████████████████████████████████████████████████████████████████| 8184.0/8184.0 [03:28<00:00, 39.21seconds/s]
100%|██████████████████████████████████████████████████████████████████████| 8184.0/8184.0 [03:40<00:00, 37.06seconds/s]
100%|██████████████████████████████████████████████████████████████████████| 8184.0/8184.0 [03:41<00:00, 36.98seconds/s]
100%|██████████████████████████████████████████████████████████████████████| 8184.0/8184.0 [03:32<00:00, 38.45seconds/s]
        Command being timed: "demucs -d cuda track6.wav"
        User time (seconds): 1004.90
        System time (seconds): 253.12
        Percent of CPU this job got: 132%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 15:46.65
        Average shared text size (kbytes): 0
        Average unshared data size (kbytes): 0
        Average stack size (kbytes): 0
        Average total size (kbytes): 0
        Maximum resident set size (kbytes): 43519832
        Average resident set size (kbytes): 0
        Major (requiring I/O) page faults: 4
        Minor (reclaiming a frame) page faults: 54473372
        Voluntary context switches: 30368
        Involuntary context switches: 2996
        Swaps: 0
        File system inputs: 0
        File system outputs: 11246632
        Socket messages sent: 0
        Socket messages received: 0
        Signals delivered: 0
        Page size (bytes): 4096
        Exit status: 0

# maximum memory usage as reported by "nvidia-smi"
|    0   N/A  N/A     30068      C   /usr/bin/python3                 5849MiB |
```

**Comment by adefossez** [comment]
Nice ! Thanks a lot for the fix :) CPU memory usage could be improved, in particular by saving to disk gradually but that will be future work.

**Comment by yuukiyuuna** [comment]
Hello, I recently discovered a new project about Prevent PyTorch's `CUDA error: out of memory`
I'm not sure if this project can solve this problem, but I think it should be a reference, maybe it can be an idea. Of course, it depends on you

https://github.com/rentruewang/koila

**Comment by adefossez** [comment]
@yuukiyuuna thanks for the pointer. However this won't help much for this project as at evaluation time we already process a single track at a time, so there is no batch size to split, so an automated method like `koila` wont be able to improve things.

**Comment by CarlGao4** [comment]
I have been trying to find a way to solve this problem. Though I didn't quite understand your code, I found that the program store the whole audio file into VRAM the moment the process starts, which requires a lot of memory though only 44 seconds of audio is used the same time if option `split` is used. I had been trying to modify your code through a whole week, but without improvement. Finally, I started a new project by rewriting IO (reading and saving file) which seems helpful. I had successfully reduced VRAM requirements to 3GB no matter the length of the audio. I'm sorry for not contributing to this project, and since the project is not finished, I haven't get it updated yet. Before I release the program, @adefossez would you mind contacting me at gaoweiqicarl1@163.com to have a discussion about the improvement?

**Comment by jayantkoolio** [comment]
Ahhhh, shit. I mixed up CPU memory with GPU memory. Thanks man. I will g5 family then, they have 24gigs of gpu memory.
